### PR TITLE
Mt/tags migration

### PR DESCRIPTION
**Pull request by maciejtreder**
## Purpose
Adding support for migrating tags

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/MigrationTool.Migration.Domain/Clients/ApiClient.cs [hunk 1]
```diff
@@ -35,12 +36,15 @@ public class ApiClient : ClientBase
...
 
+    const string AddTagRequest =
+        "{0}/subscriptions/{1}/resourceGroups/{2}/providers/Microsoft.ApiManagement/service/{3}/workspaces/{4}/apis/{5}/tags/{6}?api-version={7}";
 
 
     private readonly IApisClient ApisClient;
     private readonly IProductsClient ProductsClient;
     private readonly IApiOperationClient ApiOperationClient;
     private readonly IPolicyClient PolicyClient;
+    private readonly Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.API.Clients.Abstractions.ITagClient TagClient;
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Clients/ApiClient.cs]: lets try to keep have namespaces in using section.

File: src/MigrationTool.Migration.Domain/Clients/ProductClient.cs [hunk 2]
```diff
@@ -84,10 +95,11 @@ private async Task<Entity> CreateOrUpdateProduct(ProductApiTemplateResource reso
...
         }
 
-        public Task<IReadOnlyCollection<Entity>> FetchTags(string entityId)
+        public async Task<IReadOnlyCollection<Entity>> FetchTags(string entityId)
         {
-            // TODO: Implement
-            return Task.FromResult<IReadOnlyCollection<Entity>>(new List<Entity>());
+            var tags = await this.TagClient.GetAllTagsLinkedToProductAsync(entityId, this.ExtractorParameters);
+            return tags.ConvertAll(tag =>
+                new Entity(tag.Properties.DisplayName, EntityType.Tag, tag.Properties.DisplayName, tag));
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Clients/ProductClient.cs]: should not be
```csharp
new Entity(tag.Name, EntityType.Tag, tag.Properties.DisplayName, tag));
```
or is DisplayName unique?

File: src/MigrationTool.Migration.Domain/Executor/Operations/ApiOperationTagConnectionHandler.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    public override Type OperationType => typeof(ConnectOperation);
+
+    public override Task Handle(IMigrationOperation operation, string workspaceId)
+    {
+        var connectOperation = this.GetOperationOrThrow<ConnectOperation>(operation);
+
+        Entity apiOperation;
+        Entity tag;
+
+        this.tryGetNewEntity(connectOperation, EntityType.Tag, out tag);
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Executor/Operations/ApiOperationTagConnectionHandler.cs:31]: Add guard for for not excising tag in registry

File: src/MigrationTool.Migration.Domain/Executor/Operations/ApiTagConnectionHandler.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    public override Type OperationType => typeof(ConnectOperation);
+
+    public override Task Handle(IMigrationOperation operation, string workspaceId)
+    {
+        var connectOperation = this.GetOperationOrThrow<ConnectOperation>(operation);
+
+        Entity api;
+        Entity tag;
+
+        this.tryGetNewEntity(connectOperation, EntityType.Tag, out tag);
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Executor/Operations/ApiTagConnectionHandler.cs]: Add guard for for not excising tag in registry

File: src/MigrationTool.Migration.Domain/Executor/Operations/ProductTagConnectionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    public override Type OperationType => typeof(ConnectOperation);
+
+    public override Task Handle(IMigrationOperation operation, string workspaceId)
+    {
+        var connectOperation = this.GetOperationOrThrow<ConnectOperation>(operation);
+
+        Entity product;
+        Entity tag;
+
+        this.tryGetNewEntity(connectOperation, EntityType.Tag, out tag);
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Executor/Operations/ProductTagConnectionHandler.cs]: Add guard for for not excising tag in registry

File: src/MigrationTool.Migration.Domain/Dependencies/Resolvers/ApiOperationDependencyResolver.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+using MigrationTool.Migration.Domain.Entities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MigrationTool.Migration.Domain.Dependencies.Resolvers;
+
+public class ApiOperationDependencyResolver : IEntityDependencyResolver
```
Comment (Mielek) [inline src/MigrationTool.Migration.Domain/Dependencies/Resolvers/ApiOperationDependencyResolver.cs:12]: Need to cover case when topology sort produce something like API -> tag-> policy fragment->api operation

Comment (maciejtreder) [reply src/MigrationTool.Migration.Domain/Dependencies/Resolvers/ApiOperationDependencyResolver.cs:12]: https://github.com/Azure-Samples/api-management-workspaces-migration/issues/46

**Review by Mielek — APPROVED** [review]

**Pull request merged**